### PR TITLE
Test PRs with smoke tests

### PR DESCRIPTION
**Pull request by fg-j**
## Summary

PRs will have tests run for all latest Paketo builders for any language family subdirectories that contain changes.

The `.github/workflows/test-pull-request-<language>.yml` workflows also indirectly document which builders are compatible with which language families.

Also fixes a small bug in `smoke.sh` that caused an unbound variable error when no `--builder` was provided.

## Use Cases

1. Easily test all sample apps against the builders they should work with
1. Test PRs to ensure samples are always working

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked at least 1 issue to this pull request
* [ ] I have added an integration test, if necessary.

### CODE REVIEW

**Comment by fg-j** [comment]
I used the strategy of inspecting the builder with `pack` rather than inferring based on the builder name to open up the option for someone to have something like `testbuilder` on their Docker daemon that's a local instance of the full builder, for instance. The tests will still figure out to run the correct tests. This will be helpful for testing builders w/ sample apps directly.

**Review by ForestEckhardt — CHANGES_REQUESTED** [review]
There are still structures at the top of many of the test files that cause the file to be skipped if it is the wrong builder. Could we get rid of all of those now that we have the switch cases in `init`?

**Comment by sophiewigmore** [comment]
Should all of the smoke tests run on a PR? They take a very long time and seems like adding a sample app in Node (for example) wouldn't have an effect on the php samples. Is there a way we can just run the language family tests that were added?

**Comment by fg-j** [comment]
@ForestEckhardt I didn't realize/remember there was that skipping thing. In some ways I think that's a nicer mechanism than the switch statement in the init. Though maybe it's less explicit? Open to others' thoughts. Maybe I'll just have those skip conditionals look at the builder's run image rather than the name.

**Comment by fg-j** [comment]
@sophiewigmore These tests definitely do take a while. I can explore PR eventing to see if we can conditionally run certain tests based on files changed.

**Comment by fg-j** [comment]
@sophiewigmore I've pushed a WIP change that shows one way we could get better granularity on testing
See [this run of the Test Pull Request action](https://github.com/paketo-buildpacks/samples/pull/83/checks?check_run_id=1956837151) for an example of how the workflow runs.

**Comment by sophiewigmore** [comment]
@fg-j  Interesting. So we'd need a ` - name: Test <Language> Samples` in the workflow file for each language family. I don't _hate_ that but I agree it's a lot of duplication and set up.. Maybe it's not worth it for faster tests just for the samples repo since we're not constantly doing development on it. Interested to hear Forest's thoughts

**Comment by fg-j** [comment]
@sophiewigmore something else to consider - not particularly related but wanted to capture the context somewhere: We probably don't want to make the smoke tests a required check for the PR to merge. Here's a gist of why, in my opinion:

Suppose we use these smoke tests to test PRs and releases on the builders repos (where that check **is** required). Now suppose we release a new version of the Ruby buildpack that replaces a deprecated Ruby version with a new one. We'll want to release a builder with the new Ruby buildpack in it. We open a PR to the builder repo that updates the Ruby buildpack version. The PR tests fail because there's a sample in this repo that uses the deprecated Ruby version. So we need to change the sample app to use a supported version of Ruby. We open a PR to this repo that updates the Ruby version in the sample app. The smoke tests fail on this repo because _a version of the builder with the latest Ruby buildpack hasn't been released yet_. So we can't merge the PR on this repo, either. 

Long story short, to avoid this sort of circular dependency, I think the smoke tests should be an automatic but optional check on this repo.

**Comment by sophiewigmore** [comment]
@fg-j Thanks for the explanation. Yeah, I'm in agreement about this being optional.

**Comment by ryanmoran** [comment]
You could also break up the workflow into one per suite, then use the `paths` feature on `pull_request` events to trigger the right workflows for the PR: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths.

**Comment by fg-j** [comment]
@ryanmoran That would get us around using the imported github action, but still doesn't really address the sadness that is the duplicated TestSamples and Test<LanguageFamily> suites. Do you have thoughts on that part?

**Comment by ryanmoran** [comment]
> @ryanmoran That would get us around using the imported github action, but still doesn't really address the sadness that is the duplicated TestSamples and Test suites. Do you have thoughts on that part?

Could you colocate the tests with the samples they operate on? So, the Go tests in the `go` directory, and the PHP tests in the `php` directory. Then, in each test suite you would specify whether it runs for the given builder.

This way, you could run a single or set of suites like `go test ./go ./php` and run everything with `go test ./...`.

**Review by ForestEckhardt — APPROVED** [review]

**Review by ForestEckhardt — APPROVED** [review]

**Comment by ForestEckhardt** [comment]
@ekcasey I am going to need an approval from you to merge this one I think.

**Review by kvedurmu — APPROVED** [review]

**Comment by sophiewigmore** [comment]
@fg-j I'm not a maintainer here, but these changes look good to me!

File: java/java_test.go [hunk 1]
```diff
@@ -1,27 +1,52 @@
...
 	"testing"
+	"time"
 
 	"github.com/paketo-buildpacks/occam"
+	"github.com/paketo-buildpacks/samples/tests"
 	"github.com/sclevine/spec"
+	"github.com/sclevine/spec/report"
 
 	. "github.com/onsi/gomega"
 	. "github.com/paketo-buildpacks/occam/matchers"
```
Comment (ekcasey) [inline java/java_test.go:17]: How do you all typically group import statements in the other paketo repos. In the java buildpack repos (and cnb repos) we typically have three groups:
1. stdlib (e.g. `os`)
2. external (e.g. `github.com/paketo-buildpacks/occam`)
3. internal (e.g. `github.com/paketo-buildpacks/samples/tests`)

From this PR and other paketo buildpack repos am guessing you typically combine groups 2&3 but separate dot imports? But we don't seem to be [totally consistent about it](https://github.com/paketo-buildpacks/occam/blob/main/buildpack_store_test.go#L9-L12). No need to change anything here but maybe this is something we should add to the style guide and make consistent accross the project?

**Review by ekcasey — CHANGES_REQUESTED** [review]
I left a couple comments but the only change I think should be blocking is the go version. Other comments are suggestions for improvements that we could make separately.

This changes weren't added with this PR but I wanted to leave two additional pieces of feedback:

1. I think we should remove the application insights test for java. The app insights buildpack won't run without a binding and the current test isn't running the app insights buildpack at all.
2. I worry that asserting on the log output is a bit brittle. The built app images has [a label](https://github.com/buildpacks/spec/blob/main/platform.md#labels) that will tell you which buildpacks ran. Maybe we would rather assert on the contents of the label?

File: java/java_native_image_test.go [hunk 2]
```diff
@@ -1,18 +1,42 @@
...
 	"testing"
+	"time"
 
 	"github.com/paketo-buildpacks/occam"
+	"github.com/paketo-buildpacks/samples/tests"
 	"github.com/sclevine/spec"
+	"github.com/sclevine/spec/report"
 
 	. "github.com/onsi/gomega"
 	. "github.com/paketo-buildpacks/occam/matchers"
```
[the remark this answers is not in the document]
Comment (fg-j) [reply java/java_native_image_test.go:16]: Opened an issue to track this https://github.com/paketo-buildpacks/occam/issues/83

File: java/java_test.go (hunk 1 above)
Comment (fg-j) [reply java/java_test.go:17]: Sounds reasonable. Can you open an issue [on the community repo](https://github.com/paketo-buildpacks/community/issues) to propose a change to the style guide?

File: tests/test_utilities.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		Stderr: buffer,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	var info builderInfo
+	json.Unmarshal(buffer.Bytes(), &info)
+
+	runImage, err := reference.ParseNormalizedNamed(info.LocalInfo.RunImages[0].ImageName)
```
[the remark this answers is not in the document]
Comment (fg-j) [reply tests/test_utilities.go:54]: Now that the suites are separate, I agree that it's probably easier and clearer to handle compatibility concerns in the test runner workflows.

**Review by ekcasey — APPROVED** [review]

File: java/java_test.go (hunk 1 above)
Comment (ekcasey) [reply java/java_test.go:17]: https://github.com/paketo-buildpacks/community/issues/17
